### ISSUE TITLE
"combine(vararg flow: Flow)" method doesn't handle back pressure and skips values

### ISSUE THREAD

**Issue by n0vah** [Kotlin/kotlinx.coroutines#2671]
Method "combine(Flow, Flow, ...)" skips values if one of the input flows emits values faster than combine's flow collector process output data.
Sample to reproduce:
```kotlin
val first = MutableSharedFlow<Int>()
val second = MutableStateFlow("test")

GlobalScope.launch{
    combine(
        first,
        second
    ) { t1, t2 -> t1 to t2 }
        .collect { (first, second) ->
                Log.i("MyTag", "$second $first")
                delay(200)
        }
    }

GlobalScope.launch {
    delay(200)
    repeat(10) {
        first.emit(it + 1)
    }
}
```
Output result is:
```
MyTag: test 1
MyTag: test 3
MyTag: test 5
MyTag: test 7
MyTag: test 9
MyTag: test 10
```
logs for 2, 4, 6, 8 are missing.
[Kotlin Playground](https://play.kotlinlang.org/#eyJ2ZXJzaW9uIjoiMS4zLjcxIiwiY29kZSI6ImltcG9ydCBrb3RsaW54LmNvcm91dGluZXMuKlxuaW1wb3J0IGtvdGxpbnguY29yb3V0aW5lcy5mbG93LipcbmltcG9ydCBrb3RsaW54LmNvcm91dGluZXMuY2hhbm5lbHMuKlxuXG5mdW4gbWFpbigpIHtcblx0XHR2YWwgZmlyc3QgPSBNdXRhYmxlU2hhcmVkRmxvdzxJbnQ+KClcbiAgICAgICAgdmFsIHNlY29uZCA9IE11dGFibGVTdGF0ZUZsb3coXCJ0ZXN0XCIpXG4gICAgICAgIFxuICAgICAgICBHbG9iYWxTY29wZS5sYXVuY2h7XG4gICAgICAgICAgICBjb21iaW5lKFxuICAgICAgICAgICAgICAgIGZpcnN0LFxuICAgICAgICAgICAgICAgIHNlY29uZFxuICAgICAgICAgICAgKSB7IHQxLCB0MiAtPiB0MSB0byB0MiB9XG4gICAgICAgICAgICAgICAgLmNvbGxlY3QgeyAoZmlyc3QsIHNlY29uZCkgLT5cbiAgICAgICAgICAgICAgICAgICAgcHJpbnRsbihcIiRzZWNvbmQgJGZpcnN0XCIpXG4gICAgICAgICAgICAgICAgICAgIGRlbGF5KDEwMClcbiAgICAgICAgICAgICAgICB9XG4gICAgICAgIH1cblxuICAgICAgICBydW5CbG9ja2luZyB7XG4gICAgICAgICAgICBkZWxheSgyMDApXG4gICAgICAgICAgICByZXBlYXQoMTApIHtcbiAgICAgICAgICAgICAgICBmaXJzdC5lbWl0KGl0ICsgMSlcbiAgICAgICAgICAgIH1cbiAgICAgICAgfVxufSIsInBsYXRmb3JtIjoiamF2YSIsImFyZ3MiOiIifQ==)

I guess that is happening because inside `combineInternal` method it collects data from flows and sends them to the `resultChannel`
```kotlin
     flows[i].collect { value ->
                    resultChannel.send(Update(i, value))
                    yield() // Emulate fairness, giving each flow chance to emit
     }
``` 
but this `resultChannel` doesn't suspend anywhere to receive values. It just poll values (not suspend method)
```kotlin
 while (true) {
 ...
      element = resultChannel.poll() ?: break
      ...
```
So flow sends values to channel, but we do not receive them when 'combine` method is suspended by processing previous value in `collect` and channel doesn't wait when somebody receive his new value.

1. flows output is 1, 2, 3
2. we collect and send  "1" to the `resultChannel`
3. in `while` we poll "1" from the `resultChannel` and this value goes to the `collect`
4. While collect is processing "1" value, we send "2" to the `resultChannel`, but "combine" method is suspended by `transform` callback (waits till "collect" finish processing "1" value), so he doesn't poll  "2" value
5. we collect "3" value and sending it to the `resultChannel`
6. `collect` method finished processing "1" value,  and `combine` method is not unsuspended again
7. we poll "3" value and this value goes to the `collect`
8. As a result only "1", "3" values  went to "collect"

**Comment by dkhalanskyjb** [comment]
Yes, this is correct. According to the [documentation on `combine`](https://kotlin.github.io/kotlinx.coroutines/kotlinx-coroutines-core/kotlinx.coroutines.flow/combine.html), it emits values that are acquired by
> combining the most recently emitted values by each flow.

This method should be used to acquire a flow of the most recent information, not of every intermediate state.

Maybe [`zip`](https://kotlin.github.io/kotlinx.coroutines/kotlinx-coroutines-core/kotlinx.coroutines.flow/zip.html) would suit your use case better? It matches the values emitted by the two streams in pairs, waiting until both elements are available.

**Comment by n0vah** [comment]
@dkhalanskyjb No, "zip" won't help in my case, because zip waits for NEW values from both flows, but in my case one flow emits values sometimes (rarely), and second flow emits values often, so second flow will wait for new value from the first flow.

**Comment by dkhalanskyjb** [comment]
I believe we don't have such an operator: usually, one either wants to receive the unaltered stream of the elements, which is what `zip` does, or to acquire the most recent information, as in `combine`. What you're requesting is atypical, where outputting duplicate elements is fine, but losing some elements isn't. Could you please describe why such behavior is needed?

**Comment by n0vah** [comment]
I'm developing step counting application. Where I detect steps and emit them to the MutableSharedFlow (first flow), and I have another flow which is user's data retrieved from the database (I need user's data to calculate passed distance, and this user can be updated so I use flow). So I wanted to combine steps' and user's flows to calculate walked distance. User's flow emits data rarely  but steps' flow can emit values very often.
For now I just fixed this like this, it works fine for this case but I don't like the solution:
```
val userFlow: StateFlow<User?> = getUserFlow().stateIn(...)
val stepsFlow : SharedFlow<Int> = stepsDetector.getFlow()

stepsFlow.collect{ newSteps ->
    val user =  userFlow.value ?: return@collect
    ...
}
```

**Comment by dkhalanskyjb** [comment]
What values does `stepsDetector` emit? The number of steps walked since the last emission, so the stream looks like `1, 1, 1, 2, 2, 3, 2, 1`, right? If so, it looks like the problem here is with how the two combined streams are used asymmetrically: with `getUserFlow`, only the last emitted value is interesting and it doesn't matter if some emitted values are lost, whereas every value emitted from `stepsDetector` must be accounted for. Then, the problem can be solved by instead transforming `stepsDetector` to emit the current total number of steps walked.
```kotlin
stepsFlow
    // transform flowOf(1, 2, 3) into flowOf(0, 1, 3, 6)
    .runningFold(0) { total, newSteps -> total + newSteps }
    .combine(getUserFlow()) { totalSteps, user ->
        // do something
    }
```
This way, for both streams it won't matter if the elements are lost, and everything is clear semantically: the combination of the current up-to-date information is processed on each update.

The added bonus with this approach is that the number of accesses to the database (which are comparatively expensive) is lowered: if `stepsDetector` emits several times in quick succession, then these emissions are (very cheaply) combined.

What do you think?

**Comment by n0vah** [comment]
Unfortunately, using the total number of steps is not appropriate in my case. I need steps as they come in runtime, perhaps with folding data per few seconds (for example: first 5 second (1, 4, 1), second 5 second (3, 1) -> 6, 4).
I realised that the `combine` should work the way it works now. So I think I will close this issue and leave the implementation as I wrote in the comment earlier (Maybe with something like this `val user =  userFlow.value ?: awaitUser()`).
